_posts/2016-11-09-Rust-1.13.md

```diff
@@ -0,0 +1,469 @@
+developers to narrow commit ranges of performance regressions. It's an important
+part of our toolbox!
+
+We can use this tool to look at a [graph] of performance over the 1.13
+development cycle. This cycle covered the dates from August 16 through September
+29 (the graph begins from Augest 25th though and is filtered in a few ways to
+eliminate bogus, incomplete, or confusing results). There appear to be some big
+reductions, which are quantified on the corresponding [statistics] page.
+
+The big improvement demonstrated in the graphs, on September 1, is from an
+optimization from Niko to [cache normalized projections during
+translation][cache]. That is to say, during generation of LLVM IR, the compiler
+no longer recomputes concrete instances of associated types each time they are
+needed, but instead reuses previously-computed values. This optimization doesn't
+affect all code bases, but in code bases that exhibit certain patterns, like
+[futures-rs], where [debug mode build-time improved by up to 40%][ev1], you'll notice
+the difference.
+
+Another such optimization, that doesn't affect every crate but does affect some
+in a big way, came from Michael Woerister, and improves compile time for crates
+that export many [inline] functions. When a function is marked `#[inline]`, in
+addition to translating that function for use by the current crate, the compiler
+stores its MIR representation in the crate rlib, and translates the function to
+LLVM IR in every crate that calls it. The optimization Michael did is obvious in
+retrospect: there are some cases where inline functions are only for the
+consumption of other crates, and never called from the crate in which they are
+defined; so the compiler doesn't need to translate code for inline functions in
+the crate they are defined _unless_ they are called directly. This saves the
+cost of rustc converting the function to LLVM IR and LLVM optimizing and
+converting the function to machine code (TODO is this right about converting to
+machine code - were these functions public before?).
+
+In some cases this results in dramatic improvements. Build times for the ndarray
+crate [improved by 50%][ev2], and in the (unreleased) [winapi 0.3] crate, which
+is nothing but inline functions, rustc now generates the following LLVM IR:
+
+```ll
+; ModuleID = 'winapi.cgu-0.rs'
+source_filename = "winapi.cgu-0.rs"
+target datalayout = "e-m:w-i64:64-f80:128-n8:16:32:64-S128"
+target triple = "x86_64-pc-windows-msvc"
+
+!llvm.dbg.cu = !{!0}
+!llvm.module.flags = !{!3, !4}
+
+!0 = distinct !DICompileUnit(language: DW_LANG_Rust, file: !1, producer: "rustc version 1.13.0-dev (cf976fe2c 2016-09-15)", isOptimized: false, runtimeVersion: 0, emissionKind: FullDebug, enums: !2)
+!1 = !DIFile(filename: ".\5Csrc\5Clib.rs", directory: "C:\5Cmsys64\5Chome\5CPeter\5Cwinapi-rs")
+!2 = !{}
+!3 = !{i32 2, !"CodeView", i32 1}
+!4 = !{i32 2, !"Debug Info Version", i32 3}
```

Comment:
I picked this because it's one of the interesting examples we had on hand, and I think a little technical peek under the hood is not too risky. Maybe it could be explained better, replaced with something else visually interesting, like an inline graph, or just removed.